File: grokking_algorithms/go/substring.go
package main

import (
	"fmt"
)

func substring(a, b string) [][]int {
	cell := createMatrix(len(a), len(b))

	for i := 0; i < len(a); i++ {
		for j := 0; j < len(b); j++ {
			if a[i] == b[j] {
				if i > 0 && j > 0 {
					cell[i][j] = cell[i-1][j-1] + 1
				} else {
					cell[i][j] = 0
				}
			} else {
				cell[i][j] = 0
			}
		}
	}

	for _, i := range cell {
		fmt.Println(i)
	}
	return cell
}

File: grokking_algorithms/rust/src/substring.rs
#[allow(dead_code)]
fn substring(a: String, b: String) -> Vec<Vec<i64>> {
    let mut cell = vec![vec![0; b.len()]; a.len()];

    for i in 0..a.len() {
        for j in 0..b.len() {
            if a.chars().nth(i) == b.chars().nth(j) {
                if i > 0 && j > 0 {
                    cell[i][j] = cell[i - 1][j - 1] + 1
                } else {
                    cell[i][j] = 0
                }
            } else {
                cell[i][j] = 0
            }
        }
    }

    cell
}

#[cfg(test)]
mod tests {
    use std::vec;

    use super::*;

    #[test]
    fn test_substring_1() {
        let result = substring("fish".to_string(), "hish".to_string());

        assert_eq!(
            result,
            vec![
                vec![0, 0, 0, 0],
                vec![0, 1, 0, 0],
                vec![0, 0, 2, 0],
                vec![0, 0, 0, 3]
            ]
        )
    }
}
